lib/serializer/cbor/aws_iot_serializer_tinycbor_decoder.c

```diff
@@ -23,382 +23,453 @@
+                            * as null so user can allocate the required buffer.
+                            */
+                           cbor_value_calculate_string_length( pCborValue, &length );
+                           pDecoderObject->value.stringLength = length;
+                           returnedError = AWS_IOT_SERIALIZER_OUT_OF_MEMORY;
+                       }
+                   }
+                   else /* user provides the buffer */
+                   {
+                       length = pDecoderObject->value.stringLength;
+                       /* copy the string to the buffer */
+                       cbor_value_copy_text_string( pCborValue, ( char * ) pDecoderObject->value.pString, &length, NULL );
+                       pDecoderObject->value.stringLength = length;
+                   }
+
+                   break;
+               }
+
+            default:
+                /* Other scalar types are not supported. */
+                returnedError = AWS_IOT_SERIALIZER_NOT_SUPPORTED;
         }
     }
 
-    return returnError;
+    return returnedError;
 }
 
 /*-----------------------------------------------------------*/
 
-static AwsIotSerializerError_t _init(AwsIotSerializerDecoderObject_t * pDecoderObject,
-    uint8_t * pDataBuffer,
-    size_t maxSize)
+static AwsIotSerializerError_t _init( AwsIotSerializerDecoderObject_t * pDecoderObject,
+                                      uint8_t * pDataBuffer,
+                                      size_t maxSize )
 {
-    CborParser *pCborParser = pvPortMalloc(sizeof(CborValue));
-    _cborDecoder_t * pCborDecoder = pvPortMalloc(sizeof(_cborDecoder_t));
-    AwsIotSerializerError_t returnError = AWS_IOT_SERIALIZER_SUCCESS;
-    CborError cborError;
+    CborParser * pCborParser = pvPortMalloc( sizeof( CborParser ) );
+    _cborValueWrapper_t * pCborValueWrapper = pvPortMalloc( sizeof( _cborValueWrapper_t ) );
+
+    if( ( pCborParser == NULL ) || ( pCborValueWrapper == NULL ) )
+    {
+        vPortFree( pCborParser );
+        vPortFree( pCborValueWrapper );
+
+        return AWS_IOT_SERIALIZER_OUT_OF_MEMORY;
+    }
+
+    AwsIotSerializerError_t returnedError = AWS_IOT_SERIALIZER_SUCCESS;
+    CborError cborError = CborNoError;
 
     cborError = cbor_parser_init(
         pDataBuffer,
         maxSize,
         0,
         pCborParser,
-        &pCborDecoder->cborValue);
+        &pCborValueWrapper->cborValue );
 
-    if (_noCborError(cborError))
+    /* If init succeeds, create the decoder object. */
+    if( cborError == CborNoError )
     {
-        pCborDecoder->isOutermost = true;
-        _createDecoderObject(pCborDecoder, pDecoderObject);
+        pCborValueWrapper->isOutermost = true;
+
+        returnedError = _createDecoderObject( pCborValueWrapper, pDecoderObject );
     }
-    else
+
+    /* If there is any error or decoder object is a scalar type, free the cbor resources. */
+    if( cborError || returnedError || !_isArrayOrMap( pDecoderObject->type ) )
```

Comment:
CBOR buffer can be a single data item as well ( for example a CBOR integer ). Here we limit to either array or MAP
https://en.wikipedia.org/wiki/CBOR

Reply:
No. I free the memory of pointer of CborValue for a different reason. When it is a scalar data, the pointer of CborValue is not saved to the "handle", thus not useful. If they are not freed here, it will be memory leak.